catalog/files/impl/src/main/java/org/projectnessie/catalog/files/adls/AdlsProgrammaticOptions.java

```diff
@@ -15,117 +15,57 @@
-
-    @CanIgnoreReturnValue
-    Builder fileSystems(Map<String, ? extends AdlsFileSystemOptions> entries);
-
-    @CanIgnoreReturnValue
-    Builder putAllFileSystems(Map<String, ? extends AdlsFileSystemOptions> entries);
-
-    @CanIgnoreReturnValue
-    Builder defaultOptions(AdlsFileSystemOptions defaultOptions);
-
-    AdlsProgrammaticOptions build();
+  Map<String, AdlsNamedFileSystemOptions> fileSystems();
+
+  static AdlsProgrammaticOptions copyOf(AdlsOptions adlsOptions) {
+    ImmutableAdlsProgrammaticOptions.Builder builder =
+        ImmutableAdlsProgrammaticOptions.builder().from(adlsOptions);
+    // not copied by from() because of different type parameters in method return types
+    builder.defaultOptions(adlsOptions.defaultOptions());
+    builder.fileSystems(adlsOptions.fileSystems());
+    return builder.build();
   }
 
-  @Value.Immutable
-  interface AdlsPerFileSystemOptions extends AdlsFileSystemOptions {
-
-    static Builder builder() {
-      return ImmutableAdlsPerFileSystemOptions.builder();
+  @Value.Check
+  default AdlsProgrammaticOptions normalizeBuckets() {
+    LinkedHashMap<String, AdlsNamedFileSystemOptions> fileSystems = new LinkedHashMap<>();
+    for (String fsKey : fileSystems().keySet()) {
+      AdlsNamedFileSystemOptions options = fileSystems().get(fsKey);
+      if (options.name().isPresent()) {
+        fsKey = options.name().get();
+      } else {
+        options = ImmutableAdlsNamedFileSystemOptions.builder().from(options).name(fsKey).build();
+      }
+      if (fileSystems.put(fsKey, options) != null) {
+        throw new IllegalArgumentException("Duplicate filesystem name: " + fsKey);
+      }
     }
+    if (fileSystems.equals(fileSystems())) {
+      return this;
+    }
+    return ImmutableAdlsProgrammaticOptions.builder().from(this).fileSystems(fileSystems).build();
+  }
 
-    AdlsFileSystemOptions FALLBACK = AdlsPerFileSystemOptions.builder().build();
-
-    @SuppressWarnings("unused")
-    interface Builder {
-      @CanIgnoreReturnValue
-      Builder from(AdlsFileSystemOptions instance);
-
-      @CanIgnoreReturnValue
-      Builder account(BasicCredentials account);
-
-      @CanIgnoreReturnValue
-      Builder sasToken(KeySecret sasToken);
-
-      @CanIgnoreReturnValue
-      Builder endpoint(String endpoint);
-
-      @CanIgnoreReturnValue
-      Builder externalEndpoint(String externalEndpoint);
-
-      @CanIgnoreReturnValue
-      Builder retryPolicy(AdlsRetryStrategy retryPolicy);
-
-      @CanIgnoreReturnValue
-      Builder maxRetries(int maxRetries);
-
-      @CanIgnoreReturnValue
-      Builder tryTimeout(Duration tryTimeout);
-
-      @CanIgnoreReturnValue
-      Builder retryDelay(Duration retryDelay);
-
-      @CanIgnoreReturnValue
-      Builder maxRetryDelay(Duration maxRetryDelay);
-
-      AdlsPerFileSystemOptions build();
+  default AdlsFileSystemOptions effectiveOptionsForFileSystem(
```

Comment:
Moving this method down here instead of parent interface to make sure it is only called on a "programmatic" bean where bucket names are guaranteed to be normalized.